RuntimeLib/StellarMastery.c

```diff
@@ -0,0 +1,166 @@
+        for (Int32 Index = 0; Index < __COUNT__; Index += 1) {            \
+            MaxRandomValue += __GET_CHANCE__(&__POOL__[Index]);           \
+        }                                                                 \
+                                                                          \
+        Int32 Seed = (Int32)PlatformGetTickCount();                       \
+        Int32 RandomValue = RandomRange(&Seed, 0, MaxRandomValue);        \
+        Int32 CumulativeChance = 0;                                       \
+                                                                          \
+        for (Int32 Index = 0; Index < __COUNT__; Index += 1) {            \
+            CumulativeChance += __GET_CHANCE__(&__POOL__[Index]);         \
+            if (RandomValue < CumulativeChance) {                         \
+                __RESULT_REF__ = &__POOL__[Index];                        \
+                break;                                                    \
+            }                                                             \
+        }                                                                 \
+    }
+
+
+
+RTDataStellarLineGradeRef RTDataStellarMasteryGetLineGrade(
+	RTRuntimeRef Runtime,
+	RTDataStellarLineRef StellarLine,
+	UInt32 MaterialCount
+) {
+	RTDataStellarLineGradeRef OutGrade = NULL;
+	
+	for (UInt32 i = 0; i < StellarLine->StellarLineGradeCount; i++) {
+		RTDataStellarLineGradeRef Grade = &StellarLine->StellarLineGradeList[i];
+		if (Grade->ItemCount == MaterialCount) {
+			OutGrade = Grade;
+			break;
+		}
+	}
+	return OutGrade;
+}
+
+
+RTStellarMasterySlotRef RTCharacterStellarMasteryGetSlot(
+	RTCharacterRef Character,
+	UInt8 GroupID,
+	UInt8 SlotLine,
+	UInt8 SlotIndex
+) {
+	RTStellarMasterySlotRef OutSlot = NULL;
+
+	for (UInt32 i = 0; i < Character->Data.StellarMasteryInfo.Info.SlotCount; i++) {
+		RTStellarMasterySlotRef Slot = &Character->Data.StellarMasteryInfo.Slots[i];
+		if (Slot->GroupID == GroupID && Slot->SlotLine == SlotLine && Slot->SlotIndex == SlotIndex) {
+			OutSlot = Slot;
+			break;
+		}
+	}
+	return OutSlot;
+}
+
+RTStellarMasterySlotRef RTCharacterStellarMasteryAssertSlot(
+	RTCharacterRef Character,
+	UInt8 GroupID,
+	UInt8 SlotLine,
+	UInt8 SlotIndex
+) {
+	RTStellarMasterySlotRef OutSlot = RTCharacterStellarMasteryGetSlot(Character, GroupID, SlotLine, SlotIndex);
+	if (OutSlot == NULL) {
+		if (Character->Data.StellarMasteryInfo.Info.SlotCount < RUNTIME_CHARACTER_MAX_STELLAR_SLOT_COUNT) {
+			OutSlot = &Character->Data.StellarMasteryInfo.Slots[Character->Data.StellarMasteryInfo.Info.SlotCount];
+			
+			OutSlot->GroupID = GroupID;
+			OutSlot->SlotLine = SlotLine;
+			OutSlot->SlotIndex = SlotIndex;
+
+			OutSlot->StellarLinkGrade = 0;
+			OutSlot->StellarForceEffect = 0;
+			OutSlot->StellarForceValue = 0;
+
+			Character->Data.StellarMasteryInfo.Info.SlotCount++;
+		}
+	}
+	return OutSlot;
+}
+
+Void RTStellarMasterySetLink(
```

Comment:
Do we need this function?

Reply:
Actually no, i just used to separate business logic from controller (proc handler in our case), so i can move it back in it

Reply:
I see so when it is just to keep it out of WorldSvr then keep it its better having clean API in the runtime